python/smqtk/iqr/iqr_session.py

```diff
@@ -324,8 +361,109 @@ def reset(self):
             self._wi_seeds_used.clear()
             self.positive_descriptors.clear()
             self.negative_descriptors.clear()
-
+            self.external_positive_descriptors.clear()
+            self.external_negative_descriptors.clear()
             self.rel_index = None
             self.results = None
             self.query_f = None
             self.query_uuid = None
+    ###########################################################################
+    # I/O Methods
+
+    # I/O Constants. These should not be changed.
+    STATE_ZIP_COMPRESSION = zipfile.ZIP_DEFLATED
+    STATE_ZIP_FILENAME = "iqr_state.json"
+
+    def get_state_bytes(self,ext_pos_d):
+        """
+        Get a byte representation of the current descriptor and adjudication
+        state of this session.
+        This does not encode current results or the relevancy index's state, but
+        these can be reproduced with this state.
+        :return: State representation bytes
+        :rtype: bytes
+        """
+        def d_set_to_list(d_set):
+            # Convert set of descriptors to list of tuples:
+            #   [..., (uuid, type, vector), ...]
+            return [(d.uuid(), d.type(), d.vector().tolist()) for d in d_set]
+
+        with self:
+            # Convert session descriptors into basic values.
+            pos_d = d_set_to_list(self.positive_descriptors)
+            neg_d = d_set_to_list(self.negative_descriptors)
+            print(ext_pos_d)
+            print("self._query_uuid",self._query_uuid)
+            ext_neg_d = d_set_to_list(self.external_negative_descriptors)
+        print("external_pos2:",ext_pos_d)
+        z_buffer = io.BytesIO()
+        print("keys",ext_pos_d[self._query_uuid].type())
+        z = zipfile.ZipFile(z_buffer, 'w', self.STATE_ZIP_COMPRESSION)
+        z.writestr(self.STATE_ZIP_FILENAME, json.dumps({
+            'pos': pos_d,
+            'neg': neg_d,
+            'external_pos': [(self._query_uuid,ext_pos_d[self._query_uuid].type(),self._query_f.tolist())],
```

Comment:
Is this line maybe why you are not seeing the external positive queries reappear on state load? I'm not sure what those values being set are, but they certainly are not ``self.external_positive_descriptors``.